### PR TITLE
[tests] Refactor Redis tests to use testcontainers and run on helix

### PR DESCRIPTION
**Pull request by radical**
This moves the following tests to helix:
- Aspire.StackExchange.Redis.DistributedCaching.Tests
- Aspire.StackExchange.Redis.OutputCaching.Tests
- Aspire.StackExchange.Redis.Tests

Also,
- move the hardcoded image:tag for Redis to a shared `ContainerImageTags.cs`
- The only places with the hardcoded tags are in playground `aspire-manifest.json` files.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2856)

### CODE REVIEW

File: tests/Aspire.StackExchange.Redis.Tests/ConformanceTests.cs [hunk 1]
```diff
@@ -1,26 +1,30 @@
...
     // IConnectionMultiplexer can be created only via call to ConnectionMultiplexer.Connect
     protected override bool CanCreateClientWithoutConnectingToServer => false;
 
-    protected override bool CanConnectToServer => AspireRedisHelpers.CanConnectToServer;
+    protected override bool CanConnectToServer => RequiresDockerTheoryAttribute.IsSupported;
 
     protected override bool SupportsKeyedRegistrations => true;
 
-    protected override string[] RequiredLogCategories => ["StackExchange.Redis"];
+    protected override string[] RequiredLogCategories => ["StackExchange.Redis.ConnectionMultiplexer"];
```
Comment (radical) [inline tests/Aspire.StackExchange.Redis.Tests/ConformanceTests.cs:27]: This was changed to match what we actually get when the test is run.

File: tests/Aspire.Components.Common.Tests/ConformanceTests.cs [hunk 2]
```diff
@@ -98,6 +98,7 @@ public void OptionsTypeIsSealed()
     [InlineData(false)]
     public void HealthChecksRegistersHealthCheckService(bool enabled)
     {
+        SkipIfRequiredServerConnectionCanNotBeEstablished();
```
Comment (joperezr) [inline tests/Aspire.Components.Common.Tests/ConformanceTests.cs:101]: Assuming these were all already skipped for a different reason?

Comment (joperezr) [reply tests/Aspire.Components.Common.Tests/ConformanceTests.cs:101]: Oh no, I understand now, these were not run because they didn't have the `RunOnHelix` property.

File: Directory.Packages.props [hunk 3]
```diff
@@ -123,6 +123,7 @@
     <PackageVersion Include="Swashbuckle.AspNetCore" Version="6.5.0" />
     <PackageVersion Include="Testcontainers.RabbitMq" Version="$(TestcontainersPackageVersion)" />
     <PackageVersion Include="Testcontainers.MongoDb" Version="$(TestcontainersPackageVersion)" />
+    <PackageVersion Include="Testcontainers.Redis" Version="$(TestcontainersPackageVersion)" />
```
Comment (joperezr) [inline Directory.Packages.props:126]: Do these always version together? Or are there cases were ones may get serviced but others don't?

File: tests/Aspire.StackExchange.Redis.Tests/ConformanceTests.cs (hunk 1 above)
Comment (joperezr) [reply tests/Aspire.StackExchange.Redis.Tests/ConformanceTests.cs:27]: Interesting, how come were these not failing in the past? Or were they failing and we just weren't running them?

Comment (radical) [reply tests/Aspire.StackExchange.Redis.Tests/ConformanceTests.cs:27]: These weren't run on CI before. I don't know how they passed if they were run locally.

**Review by joperezr — APPROVED** [review]
Left few comments/questions but changes LGTM otherwise.

**Review by joperezr — COMMENTED** [review]
As an aside, we should be checking more or less what the build times are everytime we add tests to helix, just to make sure that we are not nearing the time limit, and if we are, we should consider extending it as part of the changes just to ensure that we are not introducing flakiness to the build success rate.

File: tests/Aspire.Components.Common.Tests/ConformanceTests.cs (hunk 2 above)
Comment (radical) [reply tests/Aspire.Components.Common.Tests/ConformanceTests.cs:101]: IIUC, these tests were passing on CI because they don't actually use the service. But in cases where we use a testcontainer, we get the ConnectionString from that:
https://github.com/radical/aspire/blob/3468c8513a7a9a979e322595d1da8b482482041e/tests/Aspire.StackExchange.Redis.Tests/ConformanceTests.cs#L66-L74

```csharp
    protected override void PopulateConfiguration(ConfigurationManager configuration, string? key = null)
    {
        string connectionString = RequiresDockerTheoryAttribute.IsSupported
                                    ? _containerFixture.GetConnectionString()
                                    : "localhost";
        configuration.AddInMemoryCollection([
            new KeyValuePair<string, string?>(ConformanceTests.CreateConfigKey("Aspire:StackExchange:Redis", key, "ConnectionString"), connectionString)
        ]);
```

Digging into this, the above is what redis uses so it would be fine without the `SkipIf*`. But others like `npgsql` don't, so they break. But I can have them default to `localhost` on windows too.

I'll remove these skips.

File: Directory.Packages.props (hunk 3 above)
Comment (radical) [reply Directory.Packages.props:126]: AFAICT, they seem to be versioned together.

File: tests/Aspire.Components.Common.Tests/ConformanceTests.cs (hunk 2 above)
Comment (radical) [reply tests/Aspire.Components.Common.Tests/ConformanceTests.cs:101]: fixed.

File: src/Shared/ContainerImageTags.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Aspire.Components.Common;
+
+internal static class ContainerImageTags
+{
+    public static (string image, string tag) Redis = ("redis", "7.2.4");
```
Comment (eerhardt) [inline src/Shared/ContainerImageTags.cs:8]:
```suggestion
    public static (string Image, string Tag) Redis = ("redis", "7.2.4");
```

The naming conventions are to use PascalCase because these are just like properties.

File: src/Aspire.Hosting/Aspire.Hosting.csproj [hunk 5]
```diff
@@ -24,6 +24,7 @@
     <Compile Include="$(SharedDir)StringComparers.cs" Link="Utils\StringComparers.cs" />
     <Compile Include="$(SharedDir)KnownFormats.cs" Link="Utils\KnownFormats.cs" />
     <Compile Include="$(SharedDir)CircularBuffer.cs" Link="CircularBuffer.cs" />
+    <Compile Include="$(SharedDir)ContainerImageTags.cs" Link="Utils\ContainerImageTags.cs" />
```
Comment (eerhardt) [inline src/Aspire.Hosting/Aspire.Hosting.csproj:27]: I don't think this file needs to be in the `Shared` folder. I think you can just put it top level in Aspire.Hosting, and link to there from the tests.

File: tests/Aspire.StackExchange.Redis.DistributedCaching.Tests/Aspire.StackExchange.Redis.DistributedCaching.Tests.csproj [hunk 6]
```diff
@@ -2,10 +2,12 @@
 
   <PropertyGroup>
     <TargetFramework>$(NetCurrent)</TargetFramework>
+    <RunTestsOnHelix>true</RunTestsOnHelix>
   </PropertyGroup>
 
   <ItemGroup>
     <None Include="$(RepoRoot)src\Components\Aspire.StackExchange.Redis.DistributedCaching\ConfigurationSchema.json" CopyToOutputDirectory="PreserveNewest" />
+    <Compile Include="$(RepoRoot)src\Shared\ContainerImageTags.cs" />
```
Comment (eerhardt) [inline tests/Aspire.StackExchange.Redis.DistributedCaching.Tests/Aspire.StackExchange.Redis.DistributedCaching.Tests.csproj:10]: Is this necessary? I don't see it being used in this project.

File: tests/Aspire.StackExchange.Redis.OutputCaching.Tests/Aspire.StackExchange.Redis.OutputCaching.Tests.csproj [hunk 7]
```diff
@@ -2,10 +2,12 @@
 
   <PropertyGroup>
     <TargetFramework>$(NetCurrent)</TargetFramework>
+    <RunTestsOnHelix>true</RunTestsOnHelix>
   </PropertyGroup>
 
   <ItemGroup>
     <None Include="$(RepoRoot)src\Components\Aspire.StackExchange.Redis.OutputCaching\ConfigurationSchema.json" CopyToOutputDirectory="PreserveNewest" />
+    <Compile Include="$(RepoRoot)src\Shared\ContainerImageTags.cs" />
```
Comment (eerhardt) [inline tests/Aspire.StackExchange.Redis.OutputCaching.Tests/Aspire.StackExchange.Redis.OutputCaching.Tests.csproj:10]: Is this necessary? I don't see it being used in this project.

File: tests/Aspire.StackExchange.Redis.Tests/AspireRedisExtensionsTests.cs [hunk 8]
```diff
@@ -132,44 +132,51 @@ public void ConnectionNameWinsOverConfigSection(bool useKeyed)
...
-        [false, GetConnectionString(abortConnect: true), true],
-        [true, GetConnectionString(abortConnect: false), false],
-        [false, GetConnectionString(abortConnect: false), false],
-    ];
-
-    private static IEnumerable<KeyValuePair<string, string?>> GetDefaultConfiguration() =>
-    [
-        new KeyValuePair<string, string?>("ConnectionStrings:redis", AspireRedisHelpers.TestingEndpoint)
-    ];
+    [RequiresDockerTheory]
```
Comment (eerhardt) [inline tests/Aspire.StackExchange.Redis.Tests/AspireRedisExtensionsTests.cs:140]: These tests didn't need a server before - they were just a plain `[Theory]`. Can we revert these changes and keep what was there originally?

Comment (radical) [reply tests/Aspire.StackExchange.Redis.Tests/AspireRedisExtensionsTests.cs]: Great catch, fixed.

File: tests/Aspire.StackExchange.Redis.Tests/RedisContainerFixture.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    public RedisContainer? Container { get; private set; }
+
+    public string GetConnectionString() => Container?.GetConnectionString() ??
+        throw new InvalidOperationException("The test container was not initialized.");
+
+    public async Task InitializeAsync()
+    {
+        if (RequiresDockerTheoryAttribute.IsSupported)
+        {
+            Container = new RedisBuilder()
```
Comment (mitchdenny) [inline tests/Aspire.StackExchange.Redis.Tests/RedisContainerFixture.cs:22]: Why wouldn't we use Aspire and its incoming test framework here instead of Test Containers?

Comment (eerhardt) [reply tests/Aspire.StackExchange.Redis.Tests/RedisContainerFixture.cs:22]: It's too heavyweight for our needs.

**Review by eerhardt — APPROVED** [review]
LGTM.Thanks!
